### PR TITLE
add a set -m so that stupid JVM processes that are spawned actually respond to SIGINT like good little children.

### PR DESCRIPTION
**Pull request by jkwatson**
Many props to @breedx2 for his googling & unix fu.

### CODE REVIEW

**Comment by jkwatson** [comment]
hmmm.  looks like elixir_phoenix now isn't exiting?  WTF?

**Comment by jkwatson** [comment]
I guess we stick with the `kill -9` if `-INT` doesn't work?

**Comment by jkwatson** [comment]
Not saying we want to merge this, but if we go with a plain old `kill`, then everything is happy.  What was the motivation for the `kill -INT` in the first place?

**Comment by botandrose** [comment]
Well, I was going to say that `SIGINT` tells the programs to shutdown cleanly, which results in a 0 error code, so `silently` knows all was cool, and does not display the output, where `SIGTERM` results in a nonzero error code, so we get all sorts of spew. However, looking at the Travis run, it seems to vary from impl to impl, so I'm not really sure what to think.

Okay, I just read TFA on Unix signals, and the official word is that both `SIGINT` and `SIGTERM` allow for internal cleanup before shutdown, as opposed to `SIGKILL`. In that case, I am cool with either! If `SIGTERM` as the shutdown signal is easier to support that `SIGINT`, then let's go with that. Looks like we'll just need to get both Clojure and Elixir to exit with a 0 error code after receiving SIGTERM.

What a crazy yak shave this has become! I definitely know a lot more about signals and shells than I did yesterday, which is positive. Thanks for all your help with this.

P.S. https://www.raspberrypi.org/forums/viewtopic.php?f=66&t=54219

**Comment by botandrose** [comment]
Ugh, looks like Elixir/Erlang doesn't actually support trapping signals :open_mouth:, so I don't think we can't intercept `SIGTERM` and exit cleanly. So, maybe it turns out that the current implementation is actually the best. Servers can respond to either `SIGINT` or `SIGKILL`, whichever one they like better, and we'll just try both. First `SIGINT`, then `SIGKILL`. WDYT?

**Comment by jkwatson** [comment]
IMO, the best answer is to require the server to expose a control endpoint
that will trigger a clean shutdown of the server, since it seems we can't
rely on signals.

But, that would require some sort of special auth, so maybe that's not good
either.

How about INT, then TERM then KILL, so we can at least do graceful shutdown
if at all possible for the process?

**Comment by botandrose** [comment]
Well, as it is, every impl is responding well to either `SIGINT` or `SIGTERM`, so that's good. The only remaining issue is that the `clojure_liberator` impl is returning from `SIGTERM` with a non-zero status,  resulting in spew. If we can crack that, I'd be ready to call this one good.

**Comment by jkwatson** [comment]
I'll look at that tomorrow or Wednesday.

**Comment by botandrose** [comment]
Cool, sounds good. In the meantime, I'll update the README to reflect the new state of affairs.

**Comment by jkwatson** [comment]
The tough bit here is that java is properly setting the non-zero exit code when being shutdown abormally (via SIGHALT).  It's *supposed* to return a non-zero exit code that includes as the low bytes the signal that was used to shut it down.  I think my option here is to pipe the output of the lein command to something else (like cat) that won't exit with a non-zero code.  Seems crappy, though.  :(

**Comment by botandrose** [comment]
Hmm, well we want the non-zero exit code when it fails to start, or crashes while handling an http request during runtime, etc, so that we see the error message. Maybe we'll just have to put up with some spew from this impl. :cry: Separating out the initial `bin/lein` bootstrapping into its own line would reduce the noise quite a bit on new systems, like Travis. Shrug.

I did a little research, and it seems like this may be the case of the JVM supporting the lowest common denominator, ie since Windows doesn't support POSIX signals, the JVM won't really, either. Thanks, Obama!

**Comment by jkwatson** [comment]
We can probably pipe the output into a script that will detect the exit code from SIGHALT and change it in to 0, leaving all other exit codes intact.  That might be the cleanest way to go. I'd like to figure out how to silence the lein script so that it wasn't so damn chatty about downloading dependencies.  Some quick searches didn't uncover anything super simple, though.

**Comment by botandrose** [comment]
Hey, that's an interesting idea! If you can detect the difference between the signal from a controlled exit vs any other reason, I think that would work! Clever.

Silencing the lein script's dependency gathering is what I meant by separating out the `bin/lein` bootstrapping, and let `silently` handle the rest. Something like

```shell
#!/bin/sh
set -e
. $(dirname "$0")/../../bin/silently
silently ./bin/lein deps
silently ./bin/lein trampoline run
```

As I see it, `lein deps` should handle both the initial bootstrapping of `lein` if needed, then the installation of all the dependencies, and then should return with a 0 exit code so silently will quiet it. Then the `lein trampoline run` will fire off, be killed by `SIGTERM`, and exit with an error code, but the spew will be much less. However, if you can write a shell script to filter out just the `SIGTERM` error code, then all that may not be needed!

**Comment by BartMassey** [comment]
I'm a little confused by this whole discussion. The shell has a *lot* of signal handling options. If you wouldn't mind ELI5-ing me the behavior you were seeing before and the behavior you want, I might be able to suggest something cleaner.

**Comment by botandrose** [comment]
Basically it boils down to managing output during the test suite run. If anything fails during the run, I want to see the output from process the failed, and only the output from that failing process, so we can easily determine what went wrong. If everything goes well, I don't want to see any output from any processes.

This was largely the case, until we introduced a JVM-based language into the mix, which handles signals strangely when backgrounded. It ignores `SIGINT`, forcing us to use `SIGTERM` instead. But when it exits with `SIGTERM`, it also exits with a non-zero status code, causing `silently` to think it failed, which then results in a great deal of output spew. See https://travis-ci.org/botandrose/eboshi_api_shootout/builds/86334956.

TL:DR; How can we signal the JVM to quit cleanly with a zero exit code?

**Comment by botandrose** [comment]
As an aside, `silently` is a somewhat misleading name for that script. Maybe it should be renamed to 'silent_on_success`, or something.

**Comment by jkwatson** [comment]
Looks like we have a winner! Wrapper script solves the problem

**Comment by botandrose** [comment]
I just ran this through the paces, and it seems to work exactly as advertised. Great idea! And I'm glad that the solution was ultimately encapsulated within your impl... I want to keep the test suite and the impls as decoupled as possible. Thanks for bearing with this somewhat unusual requirement!

:shipit:
